### PR TITLE
Dataset mode

### PR DESCRIPTION
**Pull request by mr-Mojo**
added dataset_mode PRED_AND_EVAL

### CODE REVIEW

File: calamari_ocr/ocr/datasets/dataset.py [hunk 1]
```diff
@@ -72,6 +73,7 @@ def run(self):
             for epoch in range(rq_params.epochs):
                 sample_idx = 0
                 if self.mode == DataSetMode.TRAIN:
+                    #no pred_and_eval bc it's shuffle
```
Comment (ChWick) [inline calamari_ocr/ocr/datasets/dataset.py]: Please add a space after comments (# no pred_and_eval ...)

File: calamari_ocr/ocr/datasets/dataset_factory.py [hunk 2]
```diff
@@ -20,6 +20,7 @@ def create_dataset(type: DataSetType,
                    non_existing_as_empty=False,
                    args: dict = None,
                    ):
+
```
Comment (ChWick) [inline calamari_ocr/ocr/datasets/dataset_factory.py]: Remove blank line (PR should not change the layout of unaffected lines)

File: calamari_ocr/ocr/datasets/file_dataset.py [hunk 3]
```diff
@@ -80,12 +80,17 @@ def __init__(self, mode: DataSetMode,
         images = [] if images is None else images
         texts = [] if texts is None else texts
 
-        if mode == DataSetMode.PREDICT:
+        #when evaluating, only texts are set via --gt argument      --> need dummy [None] imgs
```
Comment (ChWick) [inline calamari_ocr/ocr/datasets/file_dataset.py]: Same here, add a space

File: calamari_ocr/ocr/datasets/file_dataset.py [hunk 4]
```diff
@@ -80,12 +80,17 @@ def __init__(self, mode: DataSetMode,
...
+        #therefore, the and checks if images or texts is empty 
+        
+        if (mode == DataSetMode.PREDICT or mode == DataSetMode.PRED_AND_EVAL) and not texts:
             texts = [None] * len(images)
 
-        if mode == DataSetMode.EVAL:
+        if (mode == DataSetMode.EVAL or mode == DataSetMode.PRED_AND_EVAL) and not images: 
             images = [None] * len(texts)
 
+
```
Comment (ChWick) [inline calamari_ocr/ocr/datasets/file_dataset.py]: Remove this line

File: calamari_ocr/ocr/datasets/input_dataset.py [hunk 5]
```diff
@@ -202,7 +202,7 @@ def text_generator(self) -> Generator[str, None, None]:
     def generator(self, epochs=1, text_only=False) -> Generator[Tuple[np.array, List[str], Any], None, None]:
         self.check_initialized()
         for epoch in range(epochs):
-            if self.mode == DataSetMode.TRAIN:
+            if self.mode == DataSetMode.TRAIN:          #only train here, pred and eval are covered by else block
```
Comment (ChWick) [inline calamari_ocr/ocr/datasets/input_dataset.py]: space, and maybe move comment into the block (new line)

File: calamari_ocr/ocr/datasets/input_dataset.py [hunk 6]
```diff
@@ -239,7 +239,8 @@ def __init__(self,
         self.mp_context = multiprocessing.get_context('spawn')
         self.processes = processes
 
-        if data_augmenter and dataset.mode != DataSetMode.TRAIN:
+        if data_augmenter and dataset.mode != DataSetMode.TRAIN and dataset.mode != DataSetMode.PRED_AND_EVAL:
+            #no pred_and_eval bc it's augmentation
```
Comment (ChWick) [inline calamari_ocr/ocr/datasets/input_dataset.py]: Space

File: calamari_ocr/ocr/datasets/input_dataset.py [hunk 7]
```diff
@@ -322,12 +323,12 @@ def to_raw_input_dataset(self, processes=1, progress_bar=False, text_only=False)
...
         self._generate_only_non_augmented.value = prev
 
-        if self.dataset.mode == DataSetMode.TRAIN and self.data_augmentation_amount > 0:
+        if (self.dataset.mode == DataSetMode.TRAIN or self.dataset.mode == DataSetMode.PRED_AND_EVAL) and self.data_augmentation_amount > 0:
             abs_n_augs = int(self.data_augmentation_amount) if self.data_augmentation_amount >= 1 else int(self.data_augmentation_amount * len(self))
             preloaded_datas, preloaded_texts \
                 = self.data_augmenter.augment_datas(list(datas), list(texts), n_augmentations=abs_n_augs,
                                                     processes=processes, progress_bar=progress_bar)
-
+
```
Comment (ChWick) [inline calamari_ocr/ocr/datasets/input_dataset.py]: Something changed here: Remove spaces in line (?)

File: calamari_ocr/ocr/evaluator.py [hunk 8]
```diff
@@ -66,6 +66,7 @@ def run(self, _sentinel=None, gt_dataset=None, pred_dataset=None, processes=1, p
 
         if self.preloaded_gt:
             gt_data = self.preloaded_gt
+            print("GT PRELOADED")
```
Comment (ChWick) [inline calamari_ocr/ocr/evaluator.py]: Remove print. Seems to be a debug output.

File: calamari_ocr/ocr/evaluator.py [hunk 9]
```diff
@@ -80,7 +81,6 @@ def run(self, _sentinel=None, gt_dataset=None, pred_dataset=None, processes=1, p
                                                            total=len(pred_dataset),
                                                            progress_bar=progress_bar,
                                                            )]
-
```
Comment (ChWick) [inline calamari_ocr/ocr/evaluator.py]: Add the empty line you removed

File: calamari_ocr/ocr/evaluator.py [hunk 10]
```diff
@@ -207,6 +207,6 @@ def evaluate(_sentinel=None, gt_data=None, pred_data=None, processes=1, progress
 
         # evaluate single lines
         out = parallel_map(Evaluator.evaluate_single_args, [{'gt': gt, 'pred': pred, 'skip_empty_gt': skip_empty_gt} for gt, pred in zip(gt_data, pred_data)],
-                           processes=processes, progress_bar=progress_bar, desc="Evaluation")
+                           processes=1, progress_bar=progress_bar, desc="Evaluation")
```
Comment (ChWick) [inline calamari_ocr/ocr/evaluator.py]: Why did you change the processes to 1? Simply undo?

File: calamari_ocr/ocr/predictor.py [hunk 11]
```diff
@@ -262,9 +262,10 @@ def batched_data_params():
             for batch in progress_bar_wrapper(batched_data_params()):
                 sample_ids, batch_images, batch_params = zip(*batch)
                 samples = [dataset.samples()[i] for i in sample_ids]
+                current_mode = dataset.mode 
                 with ExitStack() as stack:
                     raw_dataset = [
-                        stack.enter_context(RawInputDataset(DataSetMode.PREDICT,
+                        stack.enter_context(RawInputDataset(current_mode, #DataSetMode.PREDICT,
```
Comment (ChWick) [inline calamari_ocr/ocr/predictor.py]: remove the comment, `current_mode` is fine

File: calamari_ocr/scripts/eval.py [hunk 12]
```diff
@@ -184,6 +184,7 @@ def main():
             checkpoint_params = json_format.Parse(f.read(), CheckpointParams())
             text_preproc = text_processor_from_proto(checkpoint_params.model.text_preprocessor)
 
+
```
Comment (ChWick) [inline calamari_ocr/scripts/eval.py]: Remove the line

File: calamari_ocr/scripts/eval.py [hunk 13]
```diff
@@ -199,6 +200,7 @@ def main():
         non_existing_as_empty=non_existing_as_empty,
         args={'text_index': args.pagexml_pred_text_index},
     )
+
```
Comment (ChWick) [inline calamari_ocr/scripts/eval.py]: Remove the line

File: calamari_ocr/scripts/predict.py [hunk 14]
```diff
@@ -41,13 +41,16 @@ def run(args):
 
     # load files
     input_image_files = glob_all(args.files)
+
```
Comment (ChWick) [inline calamari_ocr/scripts/predict.py]: remove the lines

File: calamari_ocr/scripts/predict.py [hunk 15]
```diff
@@ -41,13 +41,16 @@ def run(args):
...
+    
     if args.text_files:
         args.text_files = glob_all(args.text_files)
+    
 
     # skip invalid files and remove them, there wont be predictions of invalid files
     dataset = create_dataset(
         args.dataset,
-        DataSetMode.PREDICT,
+        DataSetMode.PREDICT,
```
Comment (ChWick) [inline calamari_ocr/scripts/predict.py]: Remove the "space" at the eol

File: calamari_ocr/scripts/train.py [hunk 16]
```diff
@@ -190,7 +190,7 @@ def create_train_dataset(args, dataset_args=None):
 
     dataset = create_dataset(
         args.dataset,
-        DataSetMode.TRAIN,
+        DataSetMode.TRAIN,
```
Comment (ChWick) [inline calamari_ocr/scripts/train.py]: Space at eol?

File: calamari_ocr/ocr/datasets/dataset.py (hunk 1 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/datasets/dataset.py]: done

File: calamari_ocr/ocr/datasets/dataset_factory.py (hunk 2 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/datasets/dataset_factory.py]: done

File: calamari_ocr/ocr/datasets/file_dataset.py (hunk 3 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/datasets/file_dataset.py]: done

File: calamari_ocr/ocr/datasets/file_dataset.py (hunk 4 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/datasets/file_dataset.py]: done

File: calamari_ocr/ocr/datasets/input_dataset.py (hunk 5 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/datasets/input_dataset.py]: done

File: calamari_ocr/ocr/datasets/input_dataset.py (hunk 6 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/datasets/input_dataset.py]: done

File: calamari_ocr/ocr/datasets/input_dataset.py (hunk 7 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/datasets/input_dataset.py]: done

File: calamari_ocr/ocr/evaluator.py (hunk 8 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/evaluator.py]: done

File: calamari_ocr/ocr/evaluator.py (hunk 9 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/evaluator.py]: done

File: calamari_ocr/ocr/evaluator.py (hunk 10 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/evaluator.py]: done

File: calamari_ocr/ocr/predictor.py (hunk 11 above)
Comment (mr-Mojo) [reply calamari_ocr/ocr/predictor.py]: done

File: calamari_ocr/scripts/eval.py (hunk 12 above)
Comment (mr-Mojo) [reply calamari_ocr/scripts/eval.py]: done

File: calamari_ocr/scripts/eval.py (hunk 13 above)
Comment (mr-Mojo) [reply calamari_ocr/scripts/eval.py]: done

File: calamari_ocr/scripts/predict.py (hunk 14 above)
Comment (mr-Mojo) [reply calamari_ocr/scripts/predict.py]: done

File: calamari_ocr/scripts/predict.py (hunk 15 above)
Comment (mr-Mojo) [reply calamari_ocr/scripts/predict.py]: done

File: calamari_ocr/scripts/train.py (hunk 16 above)
Comment (mr-Mojo) [reply calamari_ocr/scripts/train.py]: done

File: calamari_ocr/ocr/evaluator.py (hunk 9 above)
Comment (ChWick) [reply calamari_ocr/ocr/evaluator.py]: Now, there are spaces ;-)

File: calamari_ocr/scripts/eval.py (hunk 13 above)
Comment (ChWick) [reply calamari_ocr/scripts/eval.py]: Spaces ;-)

File: calamari_ocr/scripts/predict.py (hunk 14 above)
Comment (ChWick) [reply calamari_ocr/scripts/predict.py]: There is still a line (`git diff calamari_ocr/scripts/predict.py` should show no changes)

File: calamari_ocr/scripts/predict.py [hunk 17]
```diff
@@ -41,8 +41,10 @@ def run(args):
 
     # load files
     input_image_files = glob_all(args.files)
+    
     if args.text_files:
         args.text_files = glob_all(args.text_files)
+
```
Comment (ChWick) [inline calamari_ocr/scripts/predict.py]: This line must be removed as well
